src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -763,11 +764,31 @@ protected virtual void Dispose(bool disposing)
             // It's important that we handle all exceptions here before reporting any of them.
             // This way we can dispose all components before an error handler kicks in.
             List<Exception> exceptions = null;
+            List<ValueTask> asyncDisposables = null;
             foreach (var componentState in _componentStateById.Values)
             {
                 Log.DisposingComponent(_logger, componentState);
 
-                if (componentState.Component is IDisposable disposable)
+                // Components shouldn't need to implement IAsyncDisposable and IDisposable simultaneously,
+                // but in case they do, we prefer the async overload since we understand the sync overload
+                // is implemented for more "constrained" scenarios.
+                // Component authors are responsible for their IAsyncDisposable implementations not taking
+                // forever.
+                if (componentState.Component is IAsyncDisposable asyncDisposable)
+                {
+                    try
+                    {
+                        var task = asyncDisposable.DisposeAsync();
+                        asyncDisposables ??= new();
+                        asyncDisposables.Add(task);
```

Comment:
Should we stash these as Tasks? I'm slightly concerned that we might inadvertently violate the rules for working with `ValueTask` by deferring the await. We could avoid the allocation by checking if `DisposeAsync` succeeded synchronously:

```C#
var valueTask = asyncDisposable.DisposeAsync();
if (valueTask.IsCompletedSuccessfully) continue;

asyncDisposables ??= new();
asyncDisposables.Add(valueTask.AsTask());
```

Reply:
That's fair.

Reply:
done